### PR TITLE
Add GCC generalized atomic operation function

### PR DESCRIPTION
**Pull request by atwwww**
## 拉取/合并请求描述：(PR description)

[

#### 为什么提交这份PR (why to submit this PR)
Add GCC generalized atomic operation function

#### 你的解决方案是什么 (what is your solution)
override four gcc built-in functions as follows:
void __atomic_load (size_t size, void *mem, void *return, int model);
void __atomic_store (size_t size, void *mem, void *val, int model);
void __atomic_exchange (size_t size, void *mem, void *val, void *return, int model);
bool __atomic_compare_exchange (size_t size, void *obj, void *expected, void *desired, int success, int failure);

#### 请提供验证的bsp和config (provide the config and bsp) 

- BSP:
bsp\stm32\stm32f407-atk-explorer

- .config:
#define RT_USING_CPLUSPLUS
#define RT_USING_CPLUSPLUS11

- action:
image

]

### 当前拉取/合并请求的状态 Intent for your PR

必须选择一项 Choose one (Mandatory):

- [ ] 本拉取/合并请求是一个草稿版本 This PR is for a code-review and is intended to get feedback
- [x] 本拉取/合并请求是一个成熟版本 This PR is mature, and ready to be integrated into the repo

### 代码质量 Code Quality：

我在这个拉取/合并请求中已经考虑了 As part of this pull request, I've considered the following:

- [x] 已经仔细查看过代码改动的对比 Already check the difference between PR and old code
- [x] 代码风格正确，包括缩进空格，命名及其他风格 Style guide is adhered to, including spacing, naming and other styles
- [x] 没有垃圾代码，代码尽量精简，不包含`#if 0`代码，不包含已经被注释了的代码 All redundant code is removed and cleaned up
- [x] 所有变更均有原因及合理的，并且不会影响到其他软件组件代码或BSP All modifications are justified and not affect other components or BSP
- [x] 对难懂代码均提供对应的注释 I've commented appropriately where code is tricky
- [x] 代码是高质量的 Code in this PR is of high quality
- [x] 已经使用[formatting](https://github.com/mysterywolf/formatting) 等源码格式化工具确保格式符合[RT-Thread代码规范](https://github.com/RT-Thread/rt-thread/blob/master/documentation/contribution_guide/coding_style_cn.md) This PR complies with [RT-Thread code specification](https://github.com/RT-Thread/rt-thread/blob/master/documentation/contribution_guide/coding_style_en.md)

### CODE REVIEW

File: components/libc/cplusplus/cpp11/atomic_8.c [hunk 1]
```diff
@@ -69,6 +69,84 @@ bool __atomic_compare_exchange_8(volatile void *ptr, volatile void *expected, ui
...
+ *
+ * @param   mem is the source memory to load the value from.
+ *
+ * @param   _return is the pointer to the space where the loaded value will be stored.
+ */
+void __atomic_load(size_t size,volatile void *mem,volatile void *_return, int model)
+{
+    rt_base_t level;
+    level = rt_hw_interrupt_disable();
+    memcpy(_return, mem, size);
```
Comment (mysterywolf) [inline components/libc/cplusplus/cpp11/atomic_8.c:83]:
```suggestion
    rt_memcpy(_return, mem, size);
```

File: components/libc/cplusplus/cpp11/atomic_8.c [hunk 2]
```diff
@@ -69,6 +69,84 @@ bool __atomic_compare_exchange_8(volatile void *ptr, volatile void *expected, ui
...
+ *
+ * @param   mem is the destination memory space to store the value.
+ *
+ * @param   val is the pointer to the value to store.
+ */
+void __atomic_store(size_t size,volatile void *mem,volatile void *val, int model)
+{
+    rt_base_t level;
+    level = rt_hw_interrupt_disable();
+    memcpy(mem, val, size);
```
Comment (mysterywolf) [inline components/libc/cplusplus/cpp11/atomic_8.c:98]:
```suggestion
    rt_memcpy(mem, val, size);
```

**Review by mysterywolf — CHANGES_REQUESTED** [review]
please use rt_memcpy to replace memcpy

File: components/libc/cplusplus/cpp11/atomic_8.c (hunk 1 above)
Comment (atwwww) [reply components/libc/cplusplus/cpp11/atomic_8.c]: already corrected

File: components/libc/cplusplus/cpp11/atomic_8.c (hunk 2 above)
Comment (atwwww) [reply components/libc/cplusplus/cpp11/atomic_8.c]: already corrected in newest status

**Review by mysterywolf — APPROVED** [review]

**Pull request merged**